docs(administration): add quota period accounting docs (#224)
reanahub/docs.reana.io: docs/administration/configuration/configuring-user-quotas/index.md

!!! warning
    If the default quota values are changed, they will only be applied to newly created users. Users that already exist will not be affected by this change.

> ## Setting default periodic CPU accounting windows

Reviewer: 🤖 **Plural "windows" doesn't match the single knob being documented (LOW)**

The section discusses one configuration value (`cpu_period_reset_months`) that defines the length of *one* accounting window at a time. The plural reads slightly off.

Suggested fix:

```markdown
## Setting the default periodic CPU accounting window
```

Author: Fixed

To enable a deployment-wide default periodic CPU accounting policy for newly
created users, set the following Helm value to a positive integer:

```yaml
quota:
  cpu_period_reset_months: 3
```

Set `quota.cpu_period_reset_months` to `0` to disable the
deployment-wide default periodic CPU accounting policy.

When this value is set, each newly created user receives:

- `quota_period_months` which defines the length of the CPU accounting window
- `quota_period_start_at` which defines the start of the current active window

The current active window is derived from the user account creation time. For
example, for a user created on `2026-04-14T13:06:32` and a three-month period,
REANA will account CPU usage in windows such as `2026-04-14 -> 2026-07-14`,
`2026-07-14 -> 2026-10-14`, and so on.

As with default quota limits, changing the deployment-wide periodic CPU
setting only affects newly created users. Existing users keep their
current settings until they are updated manually.

If you want to backfill the deployment default periodic CPU policy for already
existing users, run:

```console
$ kubectl exec -i -t deployment/reana-server -- reana-db quota init-default-period

> Initialised periodic CPU reset policy for 42 existing users.

Reviewer: 🤖 **Terminology drift across Helm value, DB columns, prose, and CLI output (LOW)**

The same concept is referred to as:

- `cpu_period_reset_months` (Helm value — uses "reset")
- `quota_period_months` / `quota_period_start_at` (DB columns + REST API fields — use "period")
- "periodic CPU accounting policy" (this page's prose)
- "periodic CPU reset policy" (the CLI success message printed here on line 87)
- "CPU accounting window" / "quota period" (elsewhere)

The word "reset" appears in the Helm key and the CLI output but is never explained in the concept prose — an operator may reasonably ask whether "reset" implies usage is wiped at period boundaries (it isn't; the period just rolls over and usage is recomputed inside the new window).

Suggested fix — pick one canonical pair (e.g. *CPU accounting window* + *quota period*) and align the CLI output line and the Helm value description to it. Alternatively, add a one-sentence note acknowledging that `cpu_period_reset_months` and `quota_period_months` are the same setting at two different levels (Helm-level default vs per-user value).

Author: Fixed CLI wording

!!! warning
    If the default quota values are changed, they will only be applied to newly created users. Users that already exist will not be affected by this change.

> ## Setting default periodic CPU accounting window

Reviewer: 🤖 **Wordy section title**

Four qualifiers ("default periodic CPU accounting window") for what is essentially the default CPU quota period.

```suggestion
## Setting the default CPU quota period
```

Author: Fixed

In addition to setting CPU usage limits, you can also configure REANA to
account CPU time inside recurring quota periods instead of over the whole
lifetime of a user account. Periodic accounting is currently supported only for

> the `cpu` resource.

Reviewer: 🤖 **"account CPU time" → "track CPU time" + consistent noun**

"Account" as a transitive verb is unusual in English; "track" is the natural fit. Also unifying "periodic accounting" / "recurring quota periods" under one phrase.

```suggestion
In addition to setting CPU usage limits, you can also configure REANA to
track CPU time inside recurring quota periods instead of over the whole
lifetime of a user account. Renewable quota periods are currently supported
only for the `cpu` resource.
```

Author: Fixed

> the `cpu` resource.

Reviewer: 🤖 **Add "As of REANA 0.95 release series" version notice (supersedes earlier body rewrite for this paragraph)**

Folds the version notice into the opening sentence, in the same style as line 118 below.

```suggestion
As of REANA 0.95 release series, in addition to setting CPU usage limits, you
can configure REANA to track CPU time inside recurring quota periods instead
of over the whole lifetime of a user account. Renewable quota periods are
currently supported only for the `cpu` resource.
```

Author: Fixed

To enable a deployment-wide default periodic CPU accounting policy for newly

> created users, set the following Helm value to a positive integer:

Reviewer: 🤖 **Tighten "deployment-wide default periodic CPU accounting policy"**

Five nouns where two will do.

```suggestion
To enable a deployment-wide default CPU quota period for newly created
users, set the following Helm value to a positive integer:
```

Author: Fixed

```yaml
quota:
  cpu_period_reset_months: 3
```

Set `quota.cpu_period_reset_months` to `0` to disable the
deployment-wide default periodic CPU accounting policy. This Helm-level

> default is stored per user as `quota_period_months`.

Reviewer: 🤖 **Slightly redundant with the previous paragraph**

The "set to a positive integer" wording above already implies `0` disables. Trimming, and dropping "Helm-level default" jargon.

```suggestion
Set `quota.cpu_period_reset_months` to `0` to disable the deployment-wide
default. The chosen value is stored per user as `quota_period_months`.
```

Author: Fixed

When this value is set, each newly created user receives:

- `quota_period_months` which defines the length of the CPU accounting window

> - `quota_period_start_at` which defines the start of the current active window

Reviewer: 🤖 **"CPU accounting window" / "current active window" → "CPU quota period" / "current period"**

Matches the noun used by the user-facing docs and by the API field names.

```suggestion
- `quota_period_months` which defines the length of the CPU quota period
- `quota_period_start_at` which defines the start of the current period
```

Author: Fixed

The current active window is derived from the user account creation time. For
example, for a user created on `2026-04-14T13:06:32` and a three-month period,
REANA will account CPU usage in windows such as `2026-04-14 -> 2026-07-14`,

> `2026-07-14 -> 2026-10-14`, and so on.

Reviewer: 🤖 **Same noun standardization + `->` is unusual notation here**

Replacing the arrow with prose date ranges, which is the style used elsewhere in these docs.

```suggestion
The current period is derived from the user account creation time. For
example, for a user created on `2026-04-14T13:06:32` and a three-month period,
REANA will track CPU usage in successive periods such as `2026-04-14` to
`2026-07-14`, `2026-07-14` to `2026-10-14`, and so on.
```

Author: Fixed

As with default quota limits, changing the deployment-wide periodic CPU
setting only affects newly created users. Existing users keep their

> current settings until they are updated manually.

Reviewer: 🤖 **"deployment-wide periodic CPU setting" → "deployment-wide CPU quota period"**

Clearer and consistent.

```suggestion
As with default quota limits, changing the deployment-wide CPU quota period
only affects newly created users. Existing users keep their current
settings until they are updated manually.
```

Author: Fixed

If you want to backfill the deployment default CPU quota period for already
existing users, run:

```console
$ kubectl exec -i -t deployment/reana-server -- reana-db quota init-default-period

> Initialised the default CPU quota period for 42 existing users.

Reviewer: 🤖 **Check spelling against actual CLI output**

The rest of the REANA docs use US English ("initialize", "recalculate"); "Initialised" here stands out. This is literal CLI output, though, so the spelling should match whatever `reana-db quota init-default-period` actually prints — worth a quick check on the source side rather than a doc-only edit.

Author: Kept as is since it does return "Initialised"

```

This command only initializes users who do not already have any custom periodic

> CPU settings.

Reviewer: 🤖 **"any custom periodic CPU settings" → "a custom CPU quota period"**

Matches the noun used everywhere else.

```suggestion
This command only initializes users who do not already have a custom CPU
quota period configured.
```

Author: Fixed

Periodic quota fields updated successfully.
```

You can optionally provide the current active window start explicitly:

```console
$ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-set-period -e john.doe@example.org --resource cpu --quota-period-months 2 --quota-period-start-at 2026-05-01T00:00:00 --admin-access-token $REANA_ACCESS_TOKEN
Periodic quota fields updated successfully.
```

If you omit `--quota-period-start-at`, REANA preserves the currently stored
period start. If no period start has been stored yet, REANA initialises it from
the user account creation time and the chosen accounting cadence.

You can also force-start a new CPU quota window for a user:

```console
$ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-replenish -e john.doe@example.org --quota-period-start-at 2026-07-01T00:00:00 --admin-access-token $REANA_ACCESS_TOKEN
Quota period replenished successfully.
```

For service integrations, the same periodic CPU fields can be managed over the
quota management REST API:

```console
$ curl -X PATCH https://reana.example.org/api/quota \
    -H 'Content-Type: application/json' \

>     -H 'X-Quota-Management-Secret: $REANA_QUOTA_MANAGEMENT_SECRET' \

Reviewer: 🤖 **Single quotes prevent `$REANA_QUOTA_MANAGEMENT_SECRET` from expanding (MEDIUM)**

The curl example wraps the header in single quotes, which means bash will not expand the variable — the literal string `$REANA_QUOTA_MANAGEMENT_SECRET` will be sent as the secret value, the server will compare it against the configured secret, and the request will fail with 403. A user copy-pasting the example as-is will get a confusing failure.

The existing `GET /api/quota` example just above (lines 132-135) uses double quotes and a literal placeholder (`"X-Quota-Management-Secret: my_secret"`), which is one safe pattern. Using double quotes around the variable form is another.

Suggested fix — pick one of:

```bash
# match the GET example style (literal placeholder)
    -H "X-Quota-Management-Secret: my_secret" \

# or, if keeping the env-var form, switch the outer quotes
    -H "X-Quota-Management-Secret: $REANA_QUOTA_MANAGEMENT_SECRET" \
```

Author: Fixed!

  "usage": 12000000
}
```

## Setting individual quota periods

You can also configure user periodic CPU accounting settings. For example,
to set a two-month accounting window for one user, run:

```console

> $ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-set-period -e john.doe@example.org --resource cpu --quota-period-months 2 --admin-access-token $REANA_ACCESS_TOKEN

Reviewer: 🤖 **CLI shorthand inconsistency: `-r` vs `--resource` (LOW)**

Line 109 above uses `quota-set ... -r disk` (short form), while the new `quota-set-period` examples here use `--resource cpu` (long form). Both forms work for both commands, but the inconsistency makes the page slightly harder to skim. Pick one style — given the long lines elsewhere in this page, the long form is fine if used consistently everywhere; alternatively, switch the new examples to `-r cpu`.

Author: I’d keep this, both forms work, and I don’t think this needs a docs change.

Periodic quota fields updated successfully.
```

You can optionally provide the current active window start explicitly:

```console
$ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-set-period -e john.doe@example.org --resource cpu --quota-period-months 2 --quota-period-start-at 2026-05-01T00:00:00Z --admin-access-token $REANA_ACCESS_TOKEN
Periodic quota fields updated successfully.
```

If you omit `--quota-period-start-at`, REANA preserves the currently stored
period start. If no period start has been stored yet, REANA initialises it from
the user account creation time and the chosen accounting cadence.

You can also force-start a new CPU quota window for a user:

```console

> $ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-replenish -e john.doe@example.org --quota-period-start-at 2026-07-01T00:00:00Z --admin-access-token $REANA_ACCESS_TOKEN

Reviewer: 🤖 **`quota-replenish` is introduced without explaining how it differs from `quota-set-period --quota-period-start-at` (HIGH)**

Both commands accept `--quota-period-start-at` and both produce a new period start, so an admin reading the page side-by-side can't tell when to use one over the other. The Click help for `quota-replenish` describes it as *"Force start a new periodic quota window"* — which implies extra semantics beyond setting metadata (presumably it also resets accumulated usage, hence "replenish"). The docs should make that distinction explicit, otherwise the example reads as a redundant alternative to `quota-set-period`.

Suggested fix — add a one-sentence framing before the example:

```markdown
While `quota-set-period` updates the period metadata without touching
accumulated usage, you can also force-start a brand new CPU quota window
for a user (replenishing their quota — accumulated usage in the new
window starts at zero):
```

If the actual semantic differs from this guess, please correct accordingly — the key point is that the page should answer "why would I pick this over the previous command".

Author: Removed quota-replenish altogether

Quota period replenished successfully.
```

For service integrations, the same periodic CPU fields can be managed over the
quota management REST API:

```console
$ curl -X PATCH https://reana.example.org/api/quota \
    -H 'Content-Type: application/json' \
    -H "X-Quota-Management-Secret: $REANA_QUOTA_MANAGEMENT_SECRET" \
    --data '{
      "email": "john.doe@example.org",
      "resource_type": "cpu",
      "quota_period_months": 2,

>       "quota_period_start_at": "2026-05-01T00:00:00"

Reviewer: 🤖 **Add a `Z` timezone designator to the example timestamp (NIT)**

The PATCH request body example uses a naive `"2026-05-01T00:00:00"`. The OpenAPI spec declares `quota_period_start_at` as `format: date-time` (RFC 3339), which requires a timezone designator, and the corresponding response-side serialisation in reana-server now appends `Z`. The marshmallow parser on this PATCH input still accepts a naive string, so this works in practice — but a user copy-pasting the example as a template might also use the same naive shape when reading the response back, which would fail strict RFC 3339 validation in the Python (bravado) client.

Suggested fix — align the example with the canonical form used elsewhere in the periodic-quota docs and specs:

```json
"quota_period_start_at": "2026-05-01T00:00:00Z"
```

Author: Good catch thanks, fixed!

    }'
```

The REST API accepts one or both of `quota_period_months` and
`quota_period_start_at`.

## Refreshing quota usage after changing quota periods

Changing `quota_period_months` or `quota_period_start_at` updates the stored
period metadata immediately, but it does not by itself recompute the already
accounted CPU usage for the new window.

CPU usage is recalculated when the next quota usage update runs, either
automatically via the configured `quota.periodic_update_policy` cronjob, or

> manually via the administrative quota refresh command.

Reviewer: 🤖 **"the administrative quota refresh command" is a phantom name (LOW)**

The actual command three lines below is `reana-db quota resource-usage-update`. The phrase "the administrative quota refresh command" doesn't appear anywhere in the codebase or other docs, so a user grepping the docs for "refresh" finds nothing actionable, and the abstract label has to be re-resolved on every read.

Suggested fix — name the command directly:

```markdown
CPU usage is recalculated when the next quota usage update runs, either
automatically via the configured `quota.periodic_update_policy` cronjob, or
manually via the `reana-db quota resource-usage-update` command shown
below.
```

Author: Fixed

To force an immediate refresh, run:

```console
$ kubectl exec -i -t deployment/reana-server -- reana-db quota resource-usage-update
```

This command advances the current CPU quota window if needed and recalculates

> the user's CPU usage for the current active period only. This is particularly

Reviewer: 🤖 **"the user's CPU usage" misleads about the command's scope (HIGH)**

`reana-db quota resource-usage-update` processes *all users with active tokens* (see `update_users_cpu_quota` in `reana_db/utils.py` — the user-loop is unconditional unless an explicit user is passed, which the CLI doesn't do). The current wording ("recalculates **the user's** CPU usage for the current active period only") reads as if the command operates on a single user, which an admin might infer means there is a per-user variant or flag. There isn't.

Suggested fix:

```markdown
This command advances the current CPU quota window if needed and recalculates
each user's CPU usage for their current active period only. This is
particularly useful after changing quota period settings, when you want the
web UI and REST API responses to reflect the new accounting window
immediately.
```

(Also smoothes the "after changing... and wanting" parallel into a temporal clause.)

Author: Fixed

  "usage": 12000000
}
```

> ## Setting individual quota periods

Reviewer: 🤖 **Clarify the section is CPU-only and per-user**

The page header already says "individual quota limits" above; this section is specifically about per-user CPU quota periods.

```suggestion
## Setting per-user CPU quota periods
```

Author: Fixed

You can also configure user periodic CPU accounting settings. For example,

> to set a two-month accounting window for one user, run:

Reviewer: 🤖 **"periodic CPU accounting settings" / "accounting window" → consistent noun**

```suggestion
You can also configure CPU quota periods on a per-user basis. For example,
to set a two-month CPU quota period for one user, run:
```

Author: Fixed

> to set a two-month accounting window for one user, run:

Reviewer: 🤖 **Add "As of REANA 0.95 release series" version notice (supersedes earlier body rewrite for this paragraph)**

This section is far enough below the earlier renewable-quota section that a reader landing here directly (e.g. via search) benefits from its own version marker.

```suggestion
As of REANA 0.95 release series, you can also configure CPU quota periods on
a per-user basis. For example, to set a two-month CPU quota period for one
user, run:
```

Author: Fixed

```console
$ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-set-period -e john.doe@example.org --resource cpu --quota-period-months 2 --admin-access-token $REANA_ACCESS_TOKEN
Periodic quota fields updated successfully.
```

> You can optionally provide the current active window start explicitly:

Reviewer: 🤖 **"current active window" → "current period"**

```suggestion
You can optionally provide the current period start explicitly:
```

Author: Fixed

```console
$ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-set-period -e john.doe@example.org --resource cpu --quota-period-months 2 --quota-period-start-at 2026-05-01T00:00:00Z --admin-access-token $REANA_ACCESS_TOKEN
Periodic quota fields updated successfully.
```

If you omit `--quota-period-start-at`, REANA preserves the currently stored
period start. If no period start has been stored yet, REANA initialises it from
the user account creation time and the chosen accounting cadence.
The `--quota-period-start-at` value may also be set in the future.
Until that date is reached, CPU usage is not counted in the new periodic

> window yet, so the user effectively has unlimited quota available.

Reviewer: 🤖 **Two issues here**

1. "the chosen accounting cadence" introduces yet another term for the same thing — replace with "the configured period length".
2. "the user effectively has unlimited quota available" is not quite right: the limit still applies, but nothing is being recorded against it yet. Reframing avoids the misleading suggestion that the quota itself is unlimited.

```suggestion
If you omit `--quota-period-start-at`, REANA preserves the currently stored
period start. If no period start has been stored yet, REANA initialises it from
the user account creation time and the configured period length.
The `--quota-period-start-at` value may also be set in the future.
Until that date is reached, no CPU usage is recorded against the new period
yet.
```

Author: Fixed

To force-start a new CPU quota window for a user, reuse

> `quota-set-period` and provide only the new `--quota-period-start-at`:

Reviewer: 🤖 **"CPU quota window" → "CPU quota period"**

```suggestion
To force-start a new CPU quota period for a user, reuse
`quota-set-period` and provide only the new `--quota-period-start-at`:
```

Author: Fixed

```console
$ kubectl exec -i -t deployment/reana-server -- flask reana-admin quota-set-period -e john.doe@example.org --resource cpu --quota-period-start-at 2026-07-01T00:00:00Z --admin-access-token $REANA_ACCESS_TOKEN
Periodic quota fields updated successfully.
```

For service integrations, the same periodic CPU fields can be managed over the
quota management REST API:

```console
$ curl -X PATCH https://reana.example.org/api/quota \
    -H 'Content-Type: application/json' \
    -H "X-Quota-Management-Secret: $REANA_QUOTA_MANAGEMENT_SECRET" \
    --data '{
      "email": "john.doe@example.org",
      "resource_type": "cpu",
      "quota_period_months": 2,
      "quota_period_start_at": "2026-05-01T00:00:00Z"
    }'
```

The REST API accepts one or both of `quota_period_months` and
`quota_period_start_at`.

> ## Refreshing quota usage after changing quota periods

Reviewer: 🤖 **"Refreshing" suggests a UI action; "recalculating" is what actually happens**

Also the section deals with one period at a time, so singular reads better.

```suggestion
## Recalculating CPU usage after changing a quota period
```

Author: Fixed

Changing `quota_period_months` or `quota_period_start_at` updates the stored
period metadata immediately, but it does not by itself recompute the already

> accounted CPU usage for the new window.

Reviewer: 🤖 **"already accounted CPU usage" → "accumulated CPU usage" + consistent noun**

```suggestion
Changing `quota_period_months` or `quota_period_start_at` updates the stored
period metadata immediately, but it does not by itself recompute the
accumulated CPU usage for the new period.
```

Author: Fixed

> 
> CPU usage is recalculated when the next quota usage update runs, either
> automatically via the configured `quota.periodic_update_policy` cronjob, or
> manually via the `reana-db quota resource-usage-update` command shown
> below.
> 
> To force an immediate refresh, run:
> 
> ```console
> $ kubectl exec -i -t deployment/reana-server -- reana-db quota resource-usage-update
> ```

Reviewer: 🤖 **Trigger the existing cronjob instead of `kubectl exec`-ing the server pod**

See the top-level comment for the rationale. The job-name suffix `$(date +%s)` avoids collisions if an operator needs to trigger the refresh more than once.

````suggestion
CPU usage is recalculated when the next quota usage update runs, either
automatically on the configured `quota.periodic_update_policy` schedule, or
manually by triggering an ad-hoc run of the same cronjob.

To force an immediate refresh, spawn a one-off Job from the existing
`reana-resource-quota-update` cronjob:

```console
$ kubectl create job --from=cronjob/reana-resource-quota-update \
    reana-resource-quota-update-manual-$(date +%s)
```
````

Author: Fixed

This command advances the current CPU quota window if needed and recalculates
each user's CPU usage for their current active period only. This is
particularly useful after changing quota period settings, when you want the
web UI and REST API responses to reflect the new accounting window

> immediately.

Reviewer: 🤖 **Final consistency pass: "window" / "active period" / "accounting window" → "period"**

```suggestion
This command advances the current CPU quota period if needed and recalculates
each user's CPU usage for the current period only. This is particularly
useful after changing quota period settings, when you want the web UI and
REST API responses to reflect the new period immediately.
```

Author: Fixed
